### ISSUE TITLE
Default username validations should not prohibited spaces

### ISSUE THREAD

**Issue by MLNW** [keycloak/keycloak#34082]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

user-profile

### Describe the bug

The default user profile settings look like this in a fresh Keycloak instance in the JSON editor:

```json
{
  "attributes": [
    {
      "name": "username",
      "displayName": "${username}",
      "validations": {
        "length": {
          "min": 3,
          "max": 255
        },
        "username-prohibited-characters": {},
        "up-username-not-idn-homograph": {}
      },
      ...
    },
    ...
}
```

This suggests that there are no prohibited characters for the `username` by default.

### Version

quay.io/keycloak/keycloak:26.0.0

### Regression

- [ ] The issue is a regression

### Expected behavior

Creating a user with, e.g., spaces, should succeed.

### Actual behavior

![image](https://github.com/user-attachments/assets/82a5aaa7-b06b-4528-b8ae-338b1a6f9329)

### How to Reproduce?

Create a new realm in a fresh Keycloak instance and try to create a user with spaces.

### Anything else?

This is somewhat related to https://github.com/keycloak/keycloak/issues/28231 though I don't think there was an adequate resolution. To me the default setting should either explicitly include spaces or it is a bug that needs to be resolved. I would rather remove the implicit check for spaces instead of making it an explicitly prohibited character by default.

**Comment by MLNW** [comment]
In response to my own question at the end of the description: How to **work around** this issue by disabling this default validation?

I use the [keycloak](https://crates.io/crates/keycloak) Rust crate to interact with Keycloak. I didn't find a way to set the user profile configuration during the realm creation but I was able to modify the it after the fact:

```rust
keycloak
    .post(RealmRepresentation {
        realm: Some(KEYCLOAK_REALM.into()),
        ..Default::default()
    })
    .instrument(info_span!("Create realm"))
    .await
    .context("Failed to create realm.")?;

let mut user_profile = keycloak
    .realm_users_profile_get(KEYCLOAK_REALM)
    .instrument(info_span!("Get default user profile configuration"))
    .await
    .context("Failed to get user profile configuration.")?;

if let Some(attributes) = &mut user_profile.attributes {
    for attribute in attributes {
        if let Some(validations) = &mut attribute.validations {
            validations.remove("username-prohibited-characters");
        }
    }
}

keycloak
    .realm_users_profile_put(KEYCLOAK_REALM, user_profile)
    .instrument(info_span!("Configure user profile validation"))
    .await
    .context("Failed to configure user profile validation.")?;
```

**Comment by pedroigor** [comment]
> This suggests that there are no prohibited characters for the username by default.

@MLNW Do you mean because the validator in the JSON configuration is set like this:

```
"username-prohibited-characters": {}
```

If so, it is because the value (`{}`) is actually the validator config. See https://www.keycloak.org/docs/latest/server_admin/#_user-profile-json-configuration.

**Comment by MLNW** [comment]
> If so, it is because the value (`{}`) is actually the validator config. See https://www.keycloak.org/docs/latest/server_admin/#_user-profile-json-configuration.

Thanks for the link. Sadly, I cannot find a mention of what could be configured on this validator or even what the default configuration is. Looking for the string `username-prohibited-characters` in the link finds one occurrence. It says:

> Check if the value is a valid username as an additional barrier for attacks such as script injection. The validation is based on a default RegEx pattern that blocks characters not common in usernames. When the realm setting Email as username is enabled, this validator is skipped to allow email values.

That does not tell me how to resolve the issue.

**Comment by pedroigor** [comment]
@MLNW I see. This specific validator does not allow you to provide a different pattern. The pattern used is https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/userprofile/validator/UsernameProhibitedCharactersValidator.java#L44. Only the error message can be set as per https://www.keycloak.org/docs/latest/server_admin/#_validating-attributes.

If you think it makes sense, we can mark this issue as an enhancement so we can introduce a parameter to allow you to provide a different pattern.

**Comment by MLNW** [comment]
I'm fine with converting it to a feature. It would certainly be a good idea to improve the documentation.

I have worked around the issue sufficiently on my side (I use an email as a user name).

**Comment by kherock** [comment]
Why is it not possible to remove the `up-username-not-idn-homograph` validator altogether? This used to be possible in earlier versions of the declarative user profile feature.

**Comment by kherock** [comment]
Yeah apologies, I think I got confused by the UI. The page didn't feel like a form that used a submit button at the bottom, it seems like I had forgotten to click "Save" after removing 🤦

Consider this UX feedback 😅
